\documentclass{article}
\title{Genomic context of (differential) peaks in mouse blasts/tubes}

\usepackage[text={178mm,230mm},centering]{geometry}
\usepackage{Sweave}

\SweaveOpts{keep.source=TRUE,eps=FALSE,pdf=TRUE,width=9,height=10,prefix.string=figs/figs-mousepeaks}
\setkeys{Gin}{width=0.98\textwidth}

\begin{document}

\maketitle

\raggedright

<<setup,echo=FALSE,results=hide>>=

library(chipseq)
library(lattice)
## library(hexbin)
## library(latticeExtra)

set.seed(20081008)

@ 




<<defs,echo=FALSE>>=


logscale.components <- function(axis = c("x", "y"), base = 2)
{
    axis <- match.arg(axis)
    switch(axis,
           x = 
           function(...) {
               ans <- xscale.components.default(...)
               ans$bottom$labels$labels <- 
                   base^(ans$bottom$labels$at)
               ans
           },
           y = 
           function(...) {
               ans <- yscale.components.default(...)
               ans$left$labels$labels <- 
                   base^(ans$left$labels$at)
               ans
           })
}



summarizeLane <- function(clist, summary.fun, ..., seqlen)
{
    ## clist is a list at the lane level, with one list("+"=, "-"=) for each chromsome
    stopifnot(all(names(clist) %in% names(seqlen)))
    seqlen <- seqlen[names(clist)]
    mapply(summary.fun, clist, seqlen, ..., SIMPLIFY = FALSE)
}


summarizeReads <- 
    function(reads.list, lanes = c("1", "2", "3", "4", "6", "7", "8"), ...,
             verbose = TRUE)
{
    if (verbose) cat(paste("Processing lanes", paste(lanes, collapse = ",")), fill = TRUE)
    lapply(reads.list[lanes], summarizeLane, ...)
}


coverageSummary <- 
    function(x, max = max(end(g)) + 400L)
    ## x is a list at the lane->chromosome level, with components "+" and "-"
{
    g <- growSeqs(x)
    coverage(g, 1, max)
}


@ 




\newpage


\section*{Combining myotube and myoblast lanes}

<<>>=

load("myodMyo.rda")
library("BSgenome.Mmusculus.UCSC.mm9")
mouse.seqlens <- seqlengths(Mmusculus)

seqRanges <- lapply(myodMyo, growSeqs)
cblasts <- combineLanes(seqRanges[c("1","3","6")])
ctubes <- combineLanes(seqRanges[c("2","4","7")])

covblasts <- laneCoverage(cblasts, mouse.seqlens)
covtubes <- laneCoverage(ctubes, mouse.seqlens)

isldf <- 
    do.call(make.groups, 
            lapply(list(blasts = covblasts, 
                        tubes = covtubes,
                        control = myodMyoCoverage),
                   function(x) viewMaxs(slice(x[["chr1"]], lower = 1))))





system.time(
islandSummaries <- 
    list(blasts = do.call(make.groups, islandSummary(islands(covblasts))),
         tubes = do.call(make.groups, islandSummary(islands(covtubes))))
)

islandSummaries <- 
    lapply(islandSummaries, 
           function(x) {
               names(x)[names(x) == "which"] <- "chromosome"
               x
           })

system.time(
islandSummaries <- with(islandSummaries, make.groups(blasts, tubes))
)

depth.dist <- xtabs(~cut(maxdepth, c(0, 10, 15, 30, Inf)) + which, islandSummaries)
names(dimnames(depth.dist))[1] <- "maxdepth"
depth.dist

islandSummaries.10 <- subset(islandSummaries, maxdepth >= 10)

@ 




\newpage


\section*{Comparison between myotubes and myoblasts}


<<>>=
hexbinplot(log2(clones) ~ log2(maxdepth) | which, 
           islandSummaries.10, subset = maxdepth < 3000,
           aspect = 0.7,
           main = "islands with depth >= 10",
           type = c("g", "r"),
           xscale.components = logscale.components("x", 2),
           yscale.components = logscale.components("y", 2),
           ## scales = list(x = list(rot = 90)),
           trans = sqrt, inv = function(x) x^2)


@ 
\begin{center}
<<myoReadsByDepth,fig=TRUE,echo=FALSE,height=8>>=
plot(trellis.last.object())
@ 
\end{center}


\newpage

<<>>=

hexbinplot(log2(end-start) ~ log2(maxdepth) | which, 
           islandSummaries.10, subset = maxdepth < 3000,
           aspect = 0.7,
           main = "islands with depth >= 10",
           type = c("g", "r"),
           xscale.components = logscale.components("x", 2),
           yscale.components = logscale.components("y", 2),
           ## scales = list(x = list(rot = 90)),
           trans = sqrt, inv = function(x) x^2)


@ 
\begin{center}
<<myoDiffsByDepth,fig=TRUE,echo=FALSE,height=8>>=
plot(trellis.last.object())
@ 
\end{center}


\newpage


<<>>=

myo.ntabdf <- as.data.frame(xtabs(~round(clones) + which, islandSummaries.10))
myo.ntabdf$nreads <- as.numeric(as.character(myo.ntabdf[[1]]))
xyplot(log2(Freq) ~ log2(nreads), myo.ntabdf, 
       type = c("p", "g"), par.settings = simpleTheme(pch = 20),
       groups = which, auto.key = TRUE,
       xscale.components = logscale.components("x", 2),
       yscale.components = logscale.components("y", 2),
       scales = list(x = list(rot = 90)))

## xyplot(log2(Freq) ~ log2(nreads) | which, ntabdf, 
##        type = c("p", "g"), pch = 20,
##        xscale.components = logscale.components("x", 2),
##        yscale.components = logscale.components("y", 2),
##        scales = list(x = list(rot = 90)))



## ## all = combineLanes(list(cblasts, ctubes))

## blastIslands = islands(covblasts)

## blastIcts = readsPerIsland(blastIslands)

## blastSummaries = islandSummary(blastIslands)

@ 

\begin{center}
<<myoReadsDist,fig=TRUE,echo=FALSE,height=8>>=
plot(trellis.last.object())
@ 
\end{center}

\end{document}
